cmd/keytransparency-client/cmd/hammer.go

```diff
@@ -0,0 +1,219 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/google/keytransparency/core/authentication"
+
+	"github.com/aybabtme/uniplot/histogram"
+	"github.com/paulbellamy/ratecounter"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	tm "github.com/buger/goterm"
+	tpb "github.com/google/keytransparency/core/api/type/type_proto"
+)
+
+var (
+	maxWorkers int
+	workers    int
+	memLog     = new(bytes.Buffer)
+	ramp       time.Duration
+)
+
+func init() {
+	log.SetOutput(memLog)
+
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	RootCmd.AddCommand(hammerCmd)
+
+	hammerCmd.Flags().IntVar(&maxWorkers, "workers", 1, "Number of parallel workers")
+	hammerCmd.Flags().DurationVar(&ramp, "ramp", 1*time.Second, "Time to spend ramping up")
+}
+
+// hammerCmd represents the post command
+var hammerCmd = &cobra.Command{
+	Use:   "hammer",
+	Short: "Loadtest the server",
+	Long:  ``,
+
+	PreRun: func(cmd *cobra.Command, args []string) {
+		if err := readKeyStoreFile(); err != nil {
+			log.Fatal(err)
+		}
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		ctx := context.Background()
+		parallel(ctx, maxWorkers)
+		return nil
+	},
+}
+
+func parallel(ctx context.Context, maxWorkers int) {
+	times := make(chan time.Duration)
+	jobs := make(chan job)
+	var wg sync.WaitGroup
+	go recordLatencies(ctx, times)
+	go generateJobs(ctx, jobs)
+
+	// Slowly add workers up to maxWorkers
+	ramp := time.NewTicker(time.Duration(ramp.Nanoseconds() / int64(maxWorkers)))
+	for ; workers < maxWorkers && ctx.Err() == nil; <-ramp.C {
+		workers++
+		wg.Add(1)
+		go runJobsInThread(ctx, workers, jobs, times, &wg)
+	}
+	ramp.Stop()
+	wg.Wait()
+}
+
+type job func() error
+
+func generateJobs(ctx context.Context, jobs chan<- job) {
+	i := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case jobs <- bindJob(ctx, i):
+			i++
+		}
+	}
+}
+
+func bindJob(ctx context.Context, i int) func() error {
+	userID := fmt.Sprintf("user_%v", i)
+	return func() error {
+		return writeOp(ctx, "app1", userID)
+	}
+}
+
+func runJobsInThread(ctx context.Context, id int, jobs <-chan job, times chan<- time.Duration, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case f := <-jobs:
+			start := time.Now()
+			if err := f(); err != nil {
+				log.Printf("f(): %v", err)
+				continue
+			}
+			times <- time.Since(start)
+		}
+	}
+}
+
+func recordLatencies(ctx context.Context, times <-chan time.Duration) {
+	//start := time.Now()
+	latencies := make([]float64, 0, 1000)
+	refresh := time.NewTicker(250 * time.Millisecond)
+	newworker := time.NewTicker(time.Duration(ramp.Nanoseconds() / int64(maxWorkers)))
```

Comment:
Just worker? If not then newWorker.

Reply:
`newWorker` because it ticks at about the same time a new worker is created.